pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/SortedInvertedIndexBasedFilterOperator.java

```diff
@@ -77,20 +77,119 @@ public BaseFilterBlock nextFilterBlock(BlockId BlockId) {
+    //   clipping the ranges to [startDocId; endDocId]
+    //
+    // - "Subtractive" operators, such as NEQ and NOT IN build up a list of ranges that do not match and build a list of
+    //   matching intervals by subtracting a list of non-matching intervals from the given range of
+    //   [startDocId; endDocId]
+    //
+    // For now, we don't look at the cardinality of the column's dictionary, although we should do that if someone
+    // specifies a very large list of IN/NOT IN predicates relative to the column's cardinality or a very large/small
+    // range predicate relative to the cardinality. However, as adjacent ranges get merged before returning the final
+    // list of ranges, the only drawback is that we use a lot of memory during the filter block evaluation.
+
+    final int[] dictionaryIds;
+    boolean additiveRanges = true;
+
+    switch (predicate.getType()) {
+      case EQ:
+      case IN:
+      case RANGE:
+        dictionaryIds = evaluator.getMatchingDictionaryIds();
+        break;
+      case NEQ:
+      case NOT_IN:
+        additiveRanges = false;
+        dictionaryIds = evaluator.getNonMatchingDictionaryIds();
+        break;
+      case REGEX:
+        throw new RuntimeException("Regex is not supported");
+      default:
+        throw new RuntimeException("Unimplemented!");
+    }
+
+    if (0 < dictionaryIds.length) {
+      // Sort the dictionaryIds in ascending order, so that their respective ranges are adjacent if their
+      // dictionaryIds are adjacent
+      Arrays.sort(dictionaryIds);
+
+      IntPair lastPair = invertedIndex.getMinMaxRangeFor(dictionaryIds[0]);
+      IntRanges.clip(lastPair, startDocId, endDocId);
+
+      for (int i = 1; i < dictionaryIds.length; i++) {
+        IntPair currentPair = invertedIndex.getMinMaxRangeFor(dictionaryIds[i]);
+        IntRanges.clip(currentPair, startDocId, endDocId);
+
+        // If the previous range is degenerate, just keep the current one
+        if (IntRanges.isDegenerate(lastPair)) {
+          lastPair = currentPair;
+          continue;
+        }
+
+        // If the current range is adjacent or overlaps with the previous range, merge it into the previous range,
+        // otherwise add the previous range and keep the current one to be added
+        if (!IntRanges.rangesAreDisjoint(lastPair, currentPair)) {
+          IntRanges.mergeIntoFirst(lastPair, currentPair);
+        } else {
+          if (!IntRanges.isDegenerate(lastPair)) {
+            pairs.add(lastPair);
+          }
+          lastPair = currentPair;
+        }
+      }
+
+      // Add the last range if it's valid
+      if (!IntRanges.isDegenerate(lastPair)) {
+        pairs.add(lastPair);
+      }
+    }
+
+    if (!additiveRanges) {
+      // If the ranges are not additive ranges, our list of pairs is a list of "holes" in the [startDocId; endDocId]
+      // range. We need to take this list of pairs and invert it. To do so, there are three cases:
```

Comment:
If you call `predicate.getNonMatchingIds()` for `a <> 42`, you'll get the dictionaryId of 42, while `predicate.getMachingIds()` will return all dictionaryIds except the one for 42. The method name is very misleading.